script/dts: Generate Extended Device Tree database

erwango opened the pull request:
This PR is subset of #6762 (Codegen introduction, proposed by @b0661), focused on edts database generation.

Extended Device Tree database is a json file generated by extract_dts_includes.py script, from board device tree (dts) collated with yaml bindings.
Since it is post processed from dtc, information is displayed in an easily accessible 'dict' format:
```
 _edts dict(
   'devices':  dict(device-id :  device-struct),
   'compatibles':  dict(compatible : sorted list(device-id)),
   'device-types': dict(device-types : sorted list(compatible)),
   ...
 )

 device-struct dict(
   'device-id' : device-id,
   'compatible' : list(compatible) or compatible,
   'label' : label,
   property-name : property-value ...
 )

 device-id: opaque id for a device (do not use for other purposes),
 compatible: any of ['st,stm32-spi-fifo', ...] - 'compatible' from <binding>.yaml
 label: any of ['UART_0', 'SPI_11', ...] - label directive from DTS
 device-type: any of zephyr generic type: usb, spi, i2c, ...
```

Edit 09/11: Added device-type injection

b0661 committed:
scripts: extract_dts_includes.py: extract reg to EDTS database

Extend reg directive handling to fill the EDTS datanbase.

Signed-off-by: Bobby Noelte <b0661n0e17e@gmail.com>
Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

b0661 committed:
scripts: extract_dts_includes.py: extract interrupts to EDTS database

Extend interrupts directive handling to fill the EDTS database.

Signed-off-by: Bobby Noelte <b0661n0e17e@gmail.com>

b0661 committed:
scripts: extract_dts_includes.py: extract clocks to EDTS database

Extend clocks directive handling to fill the EDTS database.

Signed-off-by: Bobby Noelte <b0661n0e17e@gmail.com>

b0661 committed:
scripts: extract_dts_includes.py: default extract to EDTS database

Extend default directive handling to fill the EDTS database.

Signed-off-by: Bobby Noelte <b0661n0e17e@gmail.com>

b0661 committed:
scripts: extract_dts_includes.py: create EDTS database for codegen

Create extended DTS database to be used by codegen and others.

Signed-off-by: Bobby Noelte <b0661n0e17e@gmail.com>

b0661 commented on scripts/dts/edtsdatabase.py:
```diff
@@ -127,6 +135,21 @@ def _update_device_compatible(self, device_id, compatible):
+            self._edts['device-types'][device_type].append(device_id)
+            self._edts['device-types'][device_type].sort()
+
+    def insert_device_type(self, compatible, device_type):
```
Remove - already accomplished by insert.

erwango replied:
I've chosen a different device-type dict as the one you proposed.
I've chosen to display: dict(device-type : sorted list(compatible)) instead of a sorted list of device_id.
Aim was to make it easier to access to get device-type information per compatible (for DTS_ #define's generation), as well as device-type list for CI.
So I need to inject (device-type, compatible) couple, which I can't using insert.

b0661 commented on scripts/dts/extract/compatible.py:
```diff
@@ -43,6 +43,17 @@ def extract(self, node_address, yaml, prop, def_label):
+            if not isinstance(device_types, list):
+                device_types = [device_types, ]
+            for j, dev_type in enumerate(device_types):
+                edts_insert_device_type(comp, dev_type)
```
Remove - see above.

b0661 replied:
After thinking a bit - device type generation should be moved  to heuristics.py.

erwango replied:
See reasonning behind this choice above

erwango replied:
My last answer was response to first comment.

> After thinking a bit - device type generation should be moved to heuristics.py.

Yes, I hesitated indeed. Moved

b0661 commented on scripts/dts/extract/edts.py:
```diff
@@ -23,6 +23,9 @@ def edts_insert_device_property(node_address, property_path, property_value):
     device_id = edts_device_id(node_address)
     edts.insert_device_property(device_id, property_path, property_value)
 
+def edts_insert_device_type(compatible, device_type):
```
Remove - use edts_insert_device_property with 'device-type/{}'.format(device_type) instead.

erwango replied:
I'm already using edts_insert_device_property with 'device-type/{}'.format(device_type)
in heuristics class.
Aim of edts_insert_device_type is not to fill in device-type sub-node property but dedicated device-type dict. This new dict should be used at least for CI and potentially other purposes related to https://github.com/zephyrproject-rtos/zephyr/issues/9406.

erwango commented on scripts/dts/extract/gpios.py:
```diff
@@ -0,0 +1,205 @@
+
+##
+# @brief Manage gpio-x directive.
+#
```
Done, thanks

erwango commented on scripts/dts/extract/gpios.py:
```diff
@@ -0,0 +1,205 @@
+
+
+    ##
+    # @brief Extract clocks related directives
```
Done, thanks

erwango commented on scripts/dts/extract/gpios.py:
```diff
@@ -0,0 +1,205 @@
+                    .format(prop))
+
+##
+# @brief Management information for pinctrl-[x].
```
Done, Thanks

erwango commented on scripts/dts/edtsdatabase.py:
```diff
@@ -0,0 +1,217 @@
+#
```
This commit was introduced as a cut out of one of your initial commits (https://github.com/zephyrproject-rtos/zephyr/commit/1d6fa538d45a4fcc95902122b97cda195bf33511 didn't exist at that time), I take care to keep your signed-off and copyright.
Few changes have been done in order to fit with modifications done inside this PR.
Using commit https://github.com/zephyrproject-rtos/zephyr/commit/1d6fa538d45a4fcc95902122b97cda195bf33511 will require to amend it to fit current PR anyway. Am I missing something?

erwango replied:
I would really be wiling to help, but since edts review basically happens here, I do the changes in EDTS class here. Taking commits from #6762 would require for you to do the changes and then I'd need to take them, push here, see if accepted, iterate.. Process might be a bit complex and slow, don't you think?

b0661 replied:
The process I was asking for is for you to do a last time cherry-pick of  1d6fa53  into #9876. From there on I will rebase #6762 on #9876. All development will happen in #9876 and #6762 will be rebased from time to time until #9876 is merged into master.

erwango replied:
Just saw you comment, but did it before hand.
FYI I also fixed a small issue in device-types filling, so you might want to rebase again (or wait review completes)

SebastianBoe commented on scripts/dts/edtsdatabase.py:
```diff
@@ -0,0 +1,244 @@
+        for compatible in compatibles:
+            for device_id in self._edts['compatibles'].get(compatible, []):
+                device_ids[device_id] = 1
+        return list(device_ids.keys())
```
We need to re-use the code between this function and the previous.

E.g. by having a private function that returns a dict. And then have each of these two call that function and extract either the keys or the values.

galak commented:
Can you update scripts/dts/edtsdatabase.py to do something similar to scripts/dts/devicetree.py w/regards to main, and being able to run it and have it print out the database.

erwango committed:
scripts: extract_dts_includes.py: extract compatible to EDTS database

Get compatible info available by EDTS.
This way the compatible info of all activated devices can be used in
drivers.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

b0661 commented:
> Can you update scripts/dts/edtsdatabase.py to do something similar to scripts/dts/devicetree.py w/regards to main, and being able to run it and have it print out the database.

You have to run extract_dts_includes.py (not edtsdatabase.py). The "print out" will be the json file.

erwango commented:
@galak 

> Can you update scripts/dts/edtsdatabase.py to do something similar to scripts/dts/devicetree.py w/regards to main,

For now edtsdatabase.py is just a class, whole parsing/analyzing/extracting process is run by extract_dts_includes that has a main.
Instantiating edtsdatabase.py as a stand-alone runnable script would require to run extract_dts_includes.py beforehand and pass information from one to the other.

What is the goal behind your query? If point is to only generate edts.json w/o defines we can work out that by passing arguments to extract_dts_includes.py.

galak commented:
> For now edtsdatabase.py is just a class, whole parsing/analyzing/extracting process is run by extract_dts_includes that has a main.
> Instantiating edtsdatabase.py as a stand-alone runnable script would require to run extract_dts_includes.py beforehand and pass information from one to the other.
> 
> What is the goal behind your query? If point is to only generate edts.json w/o defines we can work out that by passing arguments to extract_dts_includes.py.

The input would be the edts.json file, I just wanted it to dump the data structure view of the edts.

galak commented on scripts/dts/extract/reg.py:
```diff
@@ -6,6 +6,7 @@
 
 from extract.globals import *
 from extract.directive import DTDirective
+from extract.edts import *
```
We need to add 'ranges' support.  Its not specifically related to eDTS, but will impact it.

galak commented:
Some general comments/questions:
1. Should we keep the "devices" as a tree, instead of flattened?
2. Should we keep around status = "disabled" nodes [I think we should, and have status in the tree]
3. We should have a top level 'aliases' & 'chosen' in EDTS as well.

b0661 commented:
@galak 

> The input would be the edts.json file, I just wanted it to dump the data structure view of the edts.

The json file _IS_ "the data structure view of  the edts". EDTS is a python dictionary that is written to the json file. There is no more data in EDTS, just the dictionary.

See the save function in edtsdatabase.py.

b0661 commented:
> Some general comments/questions:

 > Should we keep the "devices" as a tree, instead of flattened?

Export of a flattened database (basically a key, value list) to json does not provide any structure. Makes the json file unusable for reading. Without structure all operations must become classical select operations with some selection criteria. This is not very Pythonic and would require a more complex database engine. I´m in favour of the tree as it vastly increases readability, it is easy to understand and use, and the flattening is already implemented for cases where the flattended data is needed.

 > Should we keep around status = "disabled" nodes [I think we should, and have status in the tree]

What benefit would that bring? Info from inactive nodes is not needed by device drivers. 

> We should have a top level 'aliases' & 'chosen' in EDTS as well.

I´m not shure about  'aliases'. 'chosen' may be better feeded to Kconfig as it represents a user choice not a device characteristic.

> We should have top level "views" for GPIOs, IRQs, and clocks

No. This data can easily be generated during the device driver code generation. It is also only of interest to a specific driver. Adding all this info would move the data generation part to the extraction part. It would make extraction dependend on the usage by the drivers and would add a lot of complexity to extraction.

EDTS database is device tree info plus some info from bindings. More complex logical combinations of data shall be done by the driver (aka during code generation). There are already means to do just that (modules and templates in codegen).

galak commented:
> The json file _IS_ "the data structure view of the edts". EDTS is a python dictionary that is written to the json file. There is no more data in EDTS, just the dictionary.
> 
> See the save function in edtsdatabase.py.

Part of my reason, is this gives an example of how to load and setup EDTS code in python.  Its trivial, but useful.

galak commented:
> > Some general comments/questions:
> 
> > Should we keep the "devices" as a tree, instead of flattened?
> 
> Export of a flattened database (basically a key, value list) to json does not provide any structure. Makes the json file unusable for reading. Without structure all operations must become classical select operations with some selection criteria. This is not very Pythonic and would require a more complex database engine. I´m in favour of the tree as it vastly increases readability, it is easy to understand and use, and the flattening is already implemented for cases where the flattended data is needed.
> 

I'm not sure I follow?  Aren't we flattening the devices down to key/value pairs?  The key being "device-id"?   The tree structure I was suggesting to keep is so we could possible remove things like 'parent-device'.

For example, we have the following today:

```
{
    "devices": {
        "/soc/interrupt-controller@e000e100": {
```

Why not have this be something like:
```
{
    "devices": {
              "soc": { 
                    "interrupt-controller@e000e100": {
```
> > Should we keep around status = "disabled" nodes [I think we should, and have status in the tree]
> 
> What benefit would that bring? Info from inactive nodes is not needed by device drivers.
> 

EDTS should be useful beyond device drivers and codegen.  We might use it for validating checking, or building a cfg tool for DTS to the user.

> > We should have a top level 'aliases' & 'chosen' in EDTS as well.
> 
> I´m not shure about 'aliases'. 'chosen' may be better feeded to Kconfig as it represents a user choice not a device characteristic.

Again, don't think of EDTS limited to just codegen & zephyr.  EDTS should contain enough info for one to use it to replace the part of extract_dts_includes.py that generates the generated_dts_board.*.  So we need aliases and chosen info.

> 
> > We should have top level "views" for GPIOs, IRQs, and clocks
> 
> No. This data can easily be generated during the device driver code generation. It is also only of interest to a specific driver. Adding all this info would move the data generation part to the extraction part. It would make extraction dependend on the usage by the drivers and would add a lot of complexity to extraction.
> 

Again, you are thinking of this from a single driver or device perspective.  For example, having such views is part of how DT works and is useful.  For example, its becomes much easier to find the MAX_IRQ # for an given interrupt controller this way.

> EDTS database is device tree info plus some info from bindings. More complex logical combinations of data shall be done by the driver (aka during code generation). There are already means to do just that (modules and templates in codegen).

I think it should be more useful, to allow various tools beyond just codegen to be built on top of it.

b0661 commented:
> EDTS should be useful beyond device drivers and codegen. We might use it for validating checking, or building a cfg tool for DTS to the user.

I'm happy that the simple EDTS database class I wrote might help for a lot of use cases.

> Again, you are thinking of this from a single driver or device perspective.

Yes, I do. I wrote the EDTS database class for a use case: "Provide device tree info for compile time driver data instantiation".

> I think it should be more useful, to allow various tools beyond just codegen to be built on top of it.

Yes, definitely.

Nevertheless the extension of the EDTS database class to fit a lot more use cases should be done in future PRs. Extending the scope of the series of related PRs #9876, #6762 will question to  include them in Zephyr 1.14 even for the current use case.

@galak what you can do is to have a look on the public API of edtsdatabase.py and review whether the API contradicts your envisaged use cases. The API encapsulates the database structure, which may be changed later on if there is a need by a future use case. Please propose additions that are related to future use cases and not of use for this one in another PR/Issue.  

# About the technical things

> I'm not sure I follow? Aren't we flattening the devices down to key/value pairs? The key being "device-id"? The tree structure I was suggesting to keep is so we could possible remove things like 'parent-device'.

I had a different concept of 'flattening' than you. Yes the device tree is flattened to device specific property trees that are indexed by "device-id". "device-id" is the equivalent of a phandle in DTS. The current structure fits to the use case  "Provide device tree info for compile time driver data instantiation".

The structure you are propsing would make it much more complex to generate the data for driver data instantiation. As said - design follows the use case.

> For example, having such views is part of how DT works and is useful. For example, its becomes much easier to find the MAX_IRQ # for an given interrupt controller this way.

Views are logical combinations of data. As such a view has to process the data to create the "view". IMHO the data processing shall not be done when the database is filled but shall be done by an application that is retrieving the data from the database. There is a clear split of concepts: DTS data extraction (extract_dts_includes) and EDTS data processing (codegen or any other tool that uses the database).

The task of the database is to provide an appropriate API to fill the database and to retrieve the data from the database - thats it. The EDTS class has to be used with Python anyway. We don´t need application/ user specific processing in the database (no equivalent to stored procedures, ...). Any user of the class can use Python to create the data processing she needs.

For your example:
I don´t know whether  "MAX_IRQ #" is really necessary for a driver. If it is provided as a property it can easily be read from the database. If not you may have to search for all usages of the interrupt controller by other nodes. This **can not** be done directly at database filling (info is potentially incomplete at that time) - you need a post processing state. 
A driver specific post processing would blow up the database filling. That is why there shall be a strict separation of concepts - filling vs. usage/ processing.

> EDTS should contain enough info for one to use it to replace the part of extract_dts_includes.py that generates the generated_dts_board.*.

This may be a future extension. If done now this would trigger a another major rewrite of extract_dts_includes.py that questions inclusion in 1.14.  Clearly a candidate for another PR.

> Part of my reason, is this gives an example of how to load and setup EDTS code in python. Its trivial, but useful.

EDTS is a Python class. The json that you load will be saved identical - otherwise there would be a bug in the Python Json module.

I´m reluctant to add a main to a class library. It makes the library a tool which is not intented. It would also raise requests for other main function additions.

The usage of EDTS for testing/ prototyping is so easy - you just have to import the module. There is really no reason to add otherwise dead code to the class library. Example code should go to the documentation and not into the class file.

@erwango is responsible for this PR. Maybe he has a different opinion.

erwango committed:
scripts: edts: Fill-in 'controllers' views

Fill-in controllers hierarchy in edts database.
Aim is to provide a per controller complete hierarchized view
of the various controller types.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
scrpits/dts: edts: add alias treatment for edts generation

Add an 'aliases' dict within generated edts.json and
fill in alias information when required in device-struct.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
scripts/dts: edts: Get children at their parent's

Nodes without compatible are children of ancestor node.
We want to inject these node in a children dict of parent node.
In order to detect them more easily split get_compat function
into get_node_compat and get_parent_compat.
Then, function generate_node_definitions in extract_dts_includes
script is modified to parse parents first then children so parents
are injected in edts first and ready to welcome their children.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
scripts/dts: edts: insert chosen to edts database

Add chosen structure ti edts database

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

galak committed:
scripts: extract_dts_includes: extract gpios to EDTS database

Extract gpio property for injection in EDTS.
Cleanup no more needed function from extract_dts_includes script

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>
Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak commented on scripts/dts/extract_dts_includes.py:
```diff
@@ -349,8 +349,13 @@ def extract_property(node_compat, yaml, node_address, prop, prop_val, names,
                 for i in aliases[parent_address]:
-                    node_alias = i + '_' + def_label
-                    aliases[node_address].append(node_alias)
+                    # Build an alias name that respects device tree specs
```
Can you explain this further, I see this change the alias defines from:

I2C_1_NXP_FXOS8700_1F_BASE_ADDRESS 

to 

I2C_1_FXOS8700_1F_BASE_ADDRESS

Trying to understand what the issue was.

erwango replied:
node_alias build was of form: i2c1_NXP_FXOS8700_1F (or close to that),
and this named as injected in edts
It should rather be build using  i2c1 + "node_name" to build an close to real alias, and then converted to a define label style, so value put in edts looks like a real alias

b0661 commented on scripts/dts/extract/heuristics.py:
```diff
@@ -0,0 +1,105 @@
+# Generates in EDTS:
+# - bus/master : device id of bus master for a bus device
+# - parent-device : device id of parent device
+class DTHeuristics(DTDirective):
```
After the change of the extract function signature - why do you need to derive the class from DTDirective?

erwango replied:
Fixed

b0661 commented on scripts/dts/extract/heuristics.py:
```diff
@@ -0,0 +1,105 @@
+    # @param node_address Address of node owning the
+    #                     compatible definition.
+    # @param yaml YAML definition for the owning node.
+    # @param prop compatible property name
```
Removed from interface.

erwango replied:
Fixed

b0661 commented on scripts/dts/extract/heuristics.py:
```diff
@@ -0,0 +1,105 @@
+    #                     compatible definition.
+    # @param yaml YAML definition for the owning node.
+    # @param prop compatible property name
+    # @param def_label Define label string of node owning the
```
Removed from interface.

b0661 commented on scripts/dts/extract/heuristics.py:
```diff
@@ -0,0 +1,105 @@
+from extract.globals import *
+from extract.edts import *
+from extract.directive import DTDirective
+from extract.default import default
```
unused

erwango replied:
Fixed

b0661 commented on scripts/dts/extract/heuristics.py:
```diff
@@ -0,0 +1,105 @@
+
+from extract.globals import *
+from extract.edts import *
+from extract.directive import DTDirective
```
Do you really need this? Se following comment.

erwango replied:
Fixed

erwango commented:
@b0661 , thanks for the review, I'll be in the plane shortly, I'll get your comments early next week.

b0661 commented on scripts/dts/extract_dts_includes.py:
```diff
@@ -275,7 +275,8 @@ def extract_node_include_info(reduced, root_node_address, sub_node_address,
 
-
+    # Extract status
+    default.extract(sub_node_address, yaml, None, None)
```
You are adding a special case (prop is None) to default.extract and call it here. This is not very intuitive. 
Why not make it special by creating a status.py and calling this?

erwango replied:
Sorry, I was planning to answer you on that point, but never clicked on comment.
So, I totally agree with you this is not heuristics so removed from there.
Though, I was reluctant to create a specific class for 2 lines, and I thought default was the good class to use since we're not doing anything specific for this property, even if this require some tuning do the extract function.
In general I'd like we could avoid having people creating class each time a new property shows up. In the other hand, one can argue that it could be even worse if anyone hacks any classes w/o respect to the class.
Let me try to do this with better respect to the class:

In extract_node_include_info:
```
if 'status' in y_node['properties'].keys():
      default.extract(sub_node_address, yaml, 'status', None)
```
Then, in extract() of default class, after edts_insert_device_property(), return if def_label is none to better stick with the default class spirit.

```

b0661 commented on scripts/dts/extract_dts_includes.py:
```diff
@@ -275,6 +275,9 @@ def extract_node_include_info(reduced, root_node_address, sub_node_address,
             pass
 
+    # Extract status
+    if 'status' in y_node['properties'].keys():
```
default.extract for "status" will be called in extract_property() anyway - why does it have to be called here?

erwango replied:
There is currently no binding associated to 'status' so 'status' property is not treated by extract_property. So we could add yaml bindings for 'status' in every yaml binding. Side effect would be #define generation for each status, which is not requested today. Then it would require additional treatment to inject in edts w/o generating a define.

So I chose to add this treatment here, which is a bit off track I admit but since we're not clear yet on the use of status property, I did it the cheap way.

b0661 replied:
> There is currently no binding associated to 'status' so 'status' property ...

ok. Maybe still the "better" way to include it at least in the generic device bindings like i2c.yaml, i2c-device.yaml, uart.yaml, ... This may already cure a lot of device bindings that include the generic ones.

> Then it would require additional treatment to inject in edts w/o generating a define.

Define creation in default.py can be avoided by creating a specific extract/status.py and calling status.extract() from extract_property(). That was my original thinking when I suggested to create a status.py. 

As long as not all bindings are changed to include the status property status.extract() may be called here and in  extract_property(). The double call does not hurt and after all bindings are changed the call can be removed here. - Not really nice and definitely needs a comment to explain the removal.

erwango committed:
scripts/dts: edts: insert 'status' node property

Add 'status' in edts devices struct when available.
This operation is done as part of default class and called
before usual properties parsing since 'status' property
doesn't have binding.
Default class is modified in order not to generate #define
if def_label is set to None (hence limited to edts injection).

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
dts: bindings: Add basic bindings dedicated to dts validation

Add 'refsoc' yaml dts bindings.
These bindings will be used in a refboard dts file which will
be the reference for dts/edts sanity checks.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
boards: Add refboard dummy board

Aim of this board is to provide a reference device tree that
will be used for dts/edts scripts sanity check.
While in theory this board does not use ARM architecture and should
have its own 'ref' arch, it is set under arm/ right now in order
to avoid providing a dedicated dummy arch.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented:
Added 3 commits in order to introduce edts sanity check.

Aim is to provide sanity check to validate edts generation against a known reference.

Proposed solution compares edts.json produced from a 'reference' board dts file against inputs from a yaml reference file. This allows to exercise and test EDTSConsumer Class, but makes it painful to compare all elements (all elements have to be documented in yaml file, hence basically requests to reproduce edts to a different format).
Another way of doing would be to compare generated edts file against a reference (because reviewed and upstreamed) edts.json file. Comparison can be done using EDTSConsummer Class on both etds files, then we exercise the class but we don't fully test the class since using it on both tested and reference side. Though this solution would make much easier to validate all file elements.

tbursztyka commented on dts/bindings/clock/refsoc,clock-control.yaml:
```diff
@@ -0,0 +1,27 @@
+---
```
missing copyrights in some new yaml files

erwango replied:
Fixed.Thanks for the review!

tbursztyka reviewed (approved):
It's a big PR, but the overall is just fine.

b0661 commented on scripts/dts/edtsdatabase.py:
```diff
@@ -50,6 +99,14 @@ def get_devices_by_compatible(self, compatibles):
         return list(self._generate_device_dict(compatibles).keys())
 
+    ##
+    # @brief Get all activated compatible devices.
```
Wrong description

b0661 commented on scripts/dts/edtsdatabase.py:
```diff
@@ -50,6 +99,14 @@ def get_devices_by_compatible(self, compatibles):
+    ##
+    # @brief Get all activated compatible devices.
+    #
+    # @param compatibles compatible(s)
```
param is device id (not compatible)

erwango commented:
@b0661: fixed, thanks

galak commented:
So the device-type additions, not 100% sure what we are doing here, but thinking that type should be coming from files like dts/bindings/serial/uart.yaml.

erwango commented:
@galak :

> So the device-type additions, not 100% sure what we are doing here, but thinking that type should be coming from files like dts/bindings/serial/uart.yaml.

This is indeed the case.
`id` yaml file parameter which used to convey free (and usually redundant) text information has been changed to provide the device-type. `id` was finally renamed to `type`.
This has been done with following commit (providing the name, as SHA1 will likely change before merge)
`dts/bindings: Replace id with generic zephyr type`
`dts/bindings: Rename 'id' yaml property into 'type'`

Hence for `spi.yaml`:
```
---
title: SPI Base Structure
type: spi
version: 0.1
```

`type` should match (in lower case) the driver type provided in `drivers/<driver>/Kconfig`:
```
menuconfig SPI
	bool "SPI hardware bus support"
	help
	  Enable support for the SPI hardware bus.
```

Aim of this introduction is to be able, later on, to establish a relationship between driver symbols and device type.
This will be useful for instance in https://github.com/zephyrproject-rtos/zephyr/issues/9406 to establish that a DTS_COMPAT_XXX_SPI flag has a dependency on CONFIG_SPI

galak commented:
@erwango can we cleanup the commits a bit, a few things I noticed:
1. Merge 'scripts: edts: Sort structures when saving to json' & 'scripts/dts: edtsdatabase rework consumer class' with 'scripts/dts: Introduce edts.json generation'
2. cleanup the set of commits related to id/type

galak commented:
@erwango how do I run scripts/dts_sanity/edtscheck.py?

galak commented:
why do we need a property called device-id, seems redundant with the node name?

b0661 commented:
> why do we need a property called device-id, seems redundant with the node name?

device-id is the database key (similar to a phandle in DTS) to access device data. It is needed to easily create references to devices (as the phandle).

For ease of use and because it is unique the device-id currently is the node address. It may be optimized to another form if necessary.

galak commented:
@erwango should I expect to see 'status' in the edts.json?

galak commented:
> device-id is the database key (similar to a phandle in DTS) to access device data. It is needed to easily create references to devices (as the phandle).
> 
> For ease of use and because it is unique the device-id currently is the node address. It may be optimized to another form if necessary.

are you saying that device-id should be considered opaque than?  ie the user shouldn't make assumptions of what device-id is?

b0661 commented:
> are you saying that device-id should be considered opaque than? ie the user shouldn't make assumptions of what device-id is?

Yes it shall be considered opaque.

erwango commented:
@galak 

> @erwango how do I run scripts/dts_sanity/edtscheck.py?

python3 scripts/dts_sanity/edtscheck.py

erwango commented:
@galak :

> @erwango should I expect to see 'status' in the edts.json?

Yes. Though, it might not be present for all nodes:
Initially only 'status' value treated was 'disabled' (node discarded), and nothing was done when status was not present. So no status is equivalent to 'ok'.

erwango commented:
@galak :

> Merge 'scripts: edts: Sort structures when saving to json' & 'scripts/dts: edtsdatabase rework consumer class' with 'scripts/dts: Introduce edts.json generation'

This will be a rebase hell, lot of x-dependencies out there. I propose to split into 2 commits  EDTSProviderMixin (among the initial commits and put maximal stuff in it) and EDTSConsumerMixin (which will be among the last  commits)

> cleanup the set of commits related to id/type

Do you mean merge the 2 commits ?

erwango committed:
scripts/dts: Introduce edts.json generation

Introduce python class for extended device tree database generation
and extraction.
Extended device tree database is a json formatted database.
It contains python like dict representation of device tree information
amended with zephyr yaml bindings.

Signed-off-by: Bobby Noelte <b0661n0e17e@gmail.com>
Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

galak commented:
> Yes. Though, it might not be present for all nodes:
> Initially only 'status' value treated was 'disabled' (node discarded), and nothing was done when status was not present. So no status is equivalent to 'ok'.

we should really have status in all nodes regardless of its value.

galak commented:
> Aim of this introduction is to be able, later on, to establish a relationship between driver symbols and device type.
> This will be useful for instance in #9406 to establish that a DTS_COMPAT_XXX_SPI flag has a dependency on CONFIG_SPI

I think we should drop the type stuff for now.

erwango committed:
dts/bindings: Add yaml field 'type' for generic zephyr type

Add new 'type' field in dts yaml bindings. It is meant to
provide zephyr generic device type (i2c, spi, ...).
This information could be used for various operations that
require to have information on the generic type of a device
or a compatible (CI triage, board documentation, ...)

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
scripts: extract_dts_includes.py: extract heuristics to EDTS database

Extend directive handling to use heuristics to fill the EDTS database.

heuristics was called when treating 'compatible' property
on the assumption it is mandatory.
Actually some nodes (children) don't have 'compatible', so
move heuristics out or properties parsing and process them
just before node properties treatment.

Signed-off-by: Bobby Noelte <b0661n0e17e@gmail.com>
Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
scripts/edts: inject device-type in edts database.

Extract and inject zephyr generic device-type in edts database.
This information is extracted from dts yaml bindings 'type' field.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
scripts/dts: edts: generate a unique id for each device

Generate a unique_id with the following format for each device:
<PARENT>_<COMPAT_PARENT>_<ADDRESS_NODE>_<COMPAT_NODE_ADDRESS>

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

galak commented on scripts/dts/edtsdatabase.py:
```diff
@@ -0,0 +1,413 @@
+    #
+    # @param None
+    # @return edts 'compatibles' dict
+    def get_compatibles(self):
```
Can we remove this get_compatibles, feel like this should be kept internal

galak commented on scripts/dts/edtsdatabase.py:
```diff
@@ -0,0 +1,413 @@
+    #
+    # @param None
+    # @return edts 'aliases' dict
+    def get_aliases(self):
```
Wondering if we need to expose this or not?

galak commented on scripts/dts/edtsdatabase.py:
```diff
@@ -0,0 +1,413 @@
+    #
+    # @param None
+    # @return edts device types dict
+    def get_device_types(self):
```
lets remove, and keep internal

galak commented on scripts/dts/edtsdatabase.py:
```diff
@@ -0,0 +1,413 @@
+    #
+    # @param  None
+    # @return compatible generic device type
+    def get_controllers(self):
```
lets remove, and keep internal

galak reviewed (changes requested):
Lets remove some APIs on EDTS that I think should be internal or not needed right now

b0661 commented on scripts/dts/extract/heuristics.py:
```diff
@@ -39,10 +50,21 @@ def extract(self, node_address, yaml):
+        unique_id = self._generate_string(node_address)
+        unique_id_str = unique_id.replace("-", "_").replace(",", "_"). \
+                                          replace("/","").upper()
+        edts_insert_device_property(node_address, 'unique_id', \
```
Remove whole commit and use device-id where a unique ID for a device is needed.

erwango replied:
@b0661 , unique_id was a request from @galak , I let him answer on the need.

Why I implemented it that way:
- a)Not sure exactly on how `unique_id` will be used, so I didn't want to replace `device_id` with this `unique_id`. This is why it is implemented for now besides device_id and can be seen as redundant. I agree we can replace device_id with unique_id if we deem one of both is too many
- b)I thought about implementing a function that generate unique_id based on device_id as input. Though, still be cause I'm not sure about the future use, I chose to inject it directly in edts so i can be stored and always available afterwards, it was also easier to validate generated string was correct. I agree this could be done this way afterwards if having unique_id always available is not required
- c)I though about computing and injecting it within EDTSProviderMixin Class. Though, to generate unique_id string, it requires having access to parent information. EDTSProviderMixin class can only access information which is available in self._edts object, but at a time a node is computed, there is no guarantee its parent node is already populated. So I ended up implementing it in heuristics..
--> Btw, I think we should sort `reduced` in `generate_node_definitions`. Something like:
```
    oreduced  = collections.OrderedDict(sorted(reduced.items()))
    for k, v in oreduced.items():
```
 
Since we don't know about a) or b) yet, I think best option is to keep it that way for now.

b0661 replied:
> Since we don't know about a) or b) yet, I think best option is to keep it that way for now.

Can´t you just leave it out as long as the usage is unclear and it is not used anyway?

It should be added when the usage is clear.

erwango replied:
To be checked with @galak

galak replied:
I felt like we need this for codegen purposes.  We're going to want symbols generated that have some unique naming for a given device node.

b0661 replied:
codegen accesses the device node info by "device-id" which is unique. Unique naming (e.g. for config_info, data, ...) is done using a combination of several device properties to generate a sanitized unique device name and append specifiers like  _data, _config, ... (see devicedeclare.py).

The goal was to help people to easily identify the device symbols in the generated code and the linker reports .

```
##
# @brief Get device name
#
# Device name is generated from
# - device compatible
# - bus master address if the device is connected to a bus master
# - device address
# - parent device address if the device does not have a device address
def _device_name(device_id):
    device_name = codegen.edts().get_device_property(device_id, 'compatible/0',
						     None)
    if device_name is None:
        codegen.error("No compatible property for device id '{}'."
                      .format(device_id))

    bus_master_device_id = codegen.edts().get_device_property(device_id,
							      'bus/master', None)
    if bus_master_device_id is not None:
        reg = codegen.edts().get_device_property(bus_master_device_id, 'reg')
        try:
            # use always the first key to get first address inserted into dict
            # because reg_index may be number or name
            # reg/<reg_index>/address/<address_index> : address
            for reg_index in reg:
                for address_index in reg[reg_index]['address']:
                    bus = reg[reg_index]['address'][address_index]
                    device_name += '_' + hex(bus)[2:].zfill(8)
                    break
                break
        except:
            # this device is missing the register directive
            codegen.error("No bus master register address property for device id '{}'."
                          .format(bus_master_device_id))

    reg = codegen.edts().get_device_property(device_id, 'reg', None)
    if reg is None:
        # no reg property - take the reg property of the parent device
        parent_device_id = codegen.edts().get_device_property(
            device_id, 'parent-device', None)
        if parent_device_id:
            reg = codegen.edts().get_device_property(parent_device_id, 'reg', None)
    device_address = None
    if reg is not None:
        try:
            # use always the first key to get first address inserted into dict
            # because reg_index may be number or name
            # reg/<reg_index>/address/<address_index> : address
            for reg_index in reg:
                for address_index in reg[reg_index]['address']:
                    address = reg[reg_index]['address'][address_index]
                    device_address = hex(address)[2:].zfill(8)
                    break
                break
        except:
            # this device is missing the register directive
            pass
    if device_address is None:
        # use label instead of address
        device_address = codegen.edts().get_device_property(device_id, 'label',
                                                        '<unknown>')
        # Warn about missing reg property
        codegen.log("No register address property for device id '{}'."
                    .format(device_id), "warning", "\n")
    device_name += '_' + device_address

    device_name = device_name.replace("-", "_"). \
                              replace(",", "_"). \
                              replace(";", "_"). \
                              replace("@", "_"). \
                              replace("#", "_"). \
                              replace("&", "_"). \
                              replace("/", "_"). \
                              lower()
    return device_name
```

galak commented on scripts/dts/extract/clocks.py:
```diff
@@ -5,30 +5,73 @@
+        clock_provider_compat = get_compat(clock_provider_node_address)
+        clock_provider_bindings = yaml[clock_provider_compat]
+        clock_nr_cells = int(clock_provider['props'].get('#clock-cells', 0))
+        clock_cells_names = clock_provider_bindings.get(
```
What happens if clock_nr_cells is 4 or 10?

b0661 replied:
CELLx will be created as name - see the following code where the length of the clock_cell_names list is checked while generating clock_cell_name.

The default list could be shrinked to ['ID', ] as the other cell names will automatically be created to 'CELLx' if no name is given.

galak replied:
sounds good, that would be cleaner.

galak commented on scripts/dts/edtsdatabase.py:
```diff
@@ -76,6 +76,20 @@ def insert_device_type(self, compatible, device_type):
             self._edts['device-types'][device_type].append(compatible)
             self._edts['device-types'][device_type].sort()
 
+    def insert_device_controller(self, controller_type, controller, device_id, line):
```
This needs some work, as we kinda assume which property of the controller 'irq' or 'pin' we should "sort" the controller view by.  However, it should probably be more generic than that.  Clocks is a good example since we are sorting by some field that isn't consistent across everyone.

erwango replied:
So my understanding is that we should set `line` as a list, so we're able to inject `device_id` under `bus/bit` path or whatever `xxx\...\zzz` path is needed.
Would that be generic enough?

galak replied:
Not sure I follow, what will setting `line` as a list do?

erwango replied:
`line` is used as nesting parameter.
Today, it is a variable. hence we're constrained to nesting depth of 1 (irq, pin, bus) between controller and device:
Today we have:
for line = pin number:
```
"gpio-controller": gpioa/pin1/devicex
                                    /pin2/devicey
```
for line = irq number
```
"interrupt-controller": interupt-controllera/irq31/devicex
                                                               /irq32/devicey
```

Setting `line` as a list would allow to have a variable nesting depth and allow to have for instance:
line = bus,bit
```
"clock-controller": gpioa/busa/bit1/devicex
                                     /busa/bit2/devicey
```

So each controller could use its own nesting, making the function more generic.

galak replied:
That all makes sense, just curious how we handle 2 different clock-controllers.  Lets say one clock-controller defines one cell as 'bits', and second clock-controller defines one say as 'name'.  How will that end up working?

b0661 replied:
Different compatible will handle that.

erwango replied:
Since we give following arguments (controller_type, controller, device_id, line) we're able to have different stuff for clock-controller Y and clock-controller Z

b0661 commented on scripts/dts/edtsdatabase.py:
```diff
@@ -0,0 +1,380 @@
+    # @param device_id
+    # @return (dict)device
+    def get_device_by_device_id(self, device_id):
+        return self._edts['devices'][device_id]
```
I would prefer the device class solution that encapsulates the "raw" database scheme. It is much more versatile as functionality can be added to the retrieval of device properties.

A get_device_object_by_device_id(self, device_id) may return an object of the EDTSDevice class.

galak replied:
> I would prefer the device class solution that encapsulates the "raw" database scheme. It is much more versatile as functionality can be added to the retrieval of device properties.
> 
> EDIT:
> A get_device_object(self, device_spec) may return an object of the EDTSDevice class.

I'm good with having a device class and encapsulating the 'raw' database.

b0661 commented:
@erwango @galak 

It seems there are a lot of further ideas what to do with EDTSdatabase. Would it be possible to have a minimal implementation within this PR and make enhancements in new PRs?

 - [ ] script/dts: Generate Extended Device Tree database (minimal basis) #9876
 - [ ] scripts/dts: Extend EDTS database by generic device class
 - [ ] scripts/dts: Provide unique ID for devices in ETDS
 - [ ] scripts/dts: Extend EDTS database by clock provider device class
 - [ ] scripts/dts: Extend EDTS database by interrupt controller device class
 - [ ] ... other specific device classes based on generic device class

After merge of the minimal basis #9876 also inline code generation can be reviewed and merged:
 
  - [ ] scripts: add Zephyr inline code generation with Python #6762

b0661 reviewed (changes requested):
> scripts: edts: Fill-in 'controllers' views

The filling of the controller view is scattered over several files and will be even more scattered as other controllers are added. DTS does already have a [device]-controller directive for several devices (clocks-controller, interrupt-controller, ...).
In the orginal proposal there was a controller.py to cover the [device]-controller directives and some heuristics within heuristics.py for devices that do not have a controller directive. By this the controller view extraction could be concentrated to just 2 files - which would be in my view a real benefit for future additions and maintenance.

EDIT:
Examples of the two files:
  - https://github.com/b0661/zephyr/blob/pinctrl_test/scripts/dts/extract/controller.py
  - https://github.com/b0661/zephyr/blob/pr_pinctrl_extract/scripts/dts/extract/heuristics.py

and the addition to extract_dts_includes.py:

```
   ...
    elif prop == 'compatible':
        compatible.extract(node_address, yaml, prop, def_label)
    elif '-controller' in prop:
        controller.extract(node_address, yaml, prop, def_label)
    ...
```

erwango commented:
@b0661 

> The filling of the controller view is scattered over several files and will be even more scattered as other controllers are added. DTS does already have a [device]-controller directive for several devices (clocks-controller, interrupt-controller, ...).
In the orginal proposal there was a controller.py to cover the [device]-controller directives and some heuristics within heuristics.py for devices that do not have a controller directive. By this the controller view extraction could be concentrated to just 2 files - which would be in my view a real benefit for future additions and maintenance.

Had a detailed look to your proposal. One point is that controller information: 
- controller/provider information when inserting in edts 'devices' dict
- `edts_insert_device_controller` function requires to extract device related information ('IRQ line', 'GPIO number', 'clock cells') which currently reused from device dedicated code (gpio.py, clock.py, interrupts.py)

Moving controller related code in one single file would indeed concentrate some part but would require duplication of all these section of code (that could be rather complex) dedicated to each device in the new controller module. To my view this code duplication would not benefit to maintenance. How would you address that?

b0661 commented:
> Moving controller related code in one single file would indeed concentrate some part but would require duplication of all these section of code (that could be rather complex) dedicated to each device in the new controller module. To my view this code duplication would not benefit to maintenance. 

I was not aware that you need so much extra info.

> How would you address that?

I would make the controller view a 'View" object that computes the information from the database content (aka. View <-> Model). At computation time the database is already filled and all info like interrupts, clocks, ...  can be retrieved in a standard way.

erwango commented:
> I would make the controller view a 'View" object that computes the information from the database content (aka. View <-> Model). At computation time the database is already filled and all info like interrupts, clocks, ... can be retrieved in a standard way.

This is indeed a point that is worse discussing. I though about building the 'view' in a query rather than filling it in database. I initially also supported the fact that edts should only contain the 'device' dict.
I chose to rather fill-in the database with the 'view' for following reasons:
- Filling in 'view' when information is present will be less costly in cycles than trying to build it afterwards (with all the involved loops). This is even more true when information is queried more than once.
- As available in edts.json, it could be easily read by humans (and this could be the case also with your proposal, if view computation is done as a complementary edts building step).

This reasoning might hit some limitations if too much views are put into edts and only quite a little use of it is done afterwards. This being said, I think we still miss the experience on how edts is used. Maybe we can review internal behavior once it is deployed and used, and once we see needs for improvements.

b0661 commented on scripts/dts/edtsdatabase.py:
```diff
@@ -20,10 +20,123 @@
+    # @param property property
+    # @param property value(s)
+    # @return list of device ids matching given property value(s)
+    def get_device_ids_by_prop_value(self, property, value):
```
The function description is really misleading. I tried 
```
get_device_ids_by_prop_value('label', 'SPI_1')
```
and got an exception. 
The property is restricted to only a few; this should be described.

b0661 commented:
@erwango  you really do not know what you are doing or - worse - you do it by purpose.

You removed get_device_id_by_label() and device_properties_flattened() without a notice. This blows up any inline code generation using device_declare of any form.

I don´t know about the teamwork at your company or the Zephyr community - but for me this is the last and very obvious sign to leave you alone. I will find a way to do my work OOT.

galak commented:
> @erwango you really do not know what you are doing or - worse - you do it by purpose.
> 
> You removed get_device_id_by_label() and device_properties_flattened() without a notice. This blows up any inline code generation using device_declare of any form.
> 
> I don´t know about the teamwork at your company or the Zephyr community - but for me this is the last and very obvious sign to leave you alone. I will find a way to do my work OOT.



> It seems there are a lot of further ideas what to do with EDTSdatabase. Would it be possible to have a minimal implementation within this PR and make enhancements in new PRs?
> 
> * [ ]  script/dts: Generate Extended Device Tree database (minimal basis) #9876
> * [ ]  scripts/dts: Extend EDTS database by generic device class
> * [ ]  scripts/dts: Provide unique ID for devices in ETDS
> * [ ]  scripts/dts: Extend EDTS database by clock provider device class
> * [ ]  scripts/dts: Extend EDTS database by interrupt controller device class
> * [ ]  ... other specific device classes based on generic device class
> 
> After merge of the minimal basis #9876 also inline code generation can be reviewed and merged:
> 
> * [ ]  scripts: add Zephyr inline code generation with Python #6762

Yes, we should merge a subset to make progress.  I've unfortunately been busy with too many things, but wanted to simplify the eDTS support and just have compatible, regs, irqs, and basic properties merged.

erwango commented:
@b0661 

> You removed get_device_id_by_label() and device_properties_flattened() without a notice. This blows up any inline code generation using device_declare of any form.
I don´t know about the teamwork at your company or the Zephyr community - but for me this is the last and very obvious sign to leave you alone. I will find a way to do my work OOT

I acknowledge your remark. I was trying to follow recommendations for merge, but doing so I lost connection with codegen work.

@galak 

> Yes, we should merge a subset to make progress. I've unfortunately been busy with too many things, but wanted to simplify the eDTS support 

ok

> and just have compatible, regs, irqs, and basic properties merged.

I think @b0661 is right here. We should push something coherent from codegen point of view, since this is the initial target. Ok to remove controllers/device-type/status/chosen. But we need to keep compatibles and full device dict (with reworked reg if needed).

erwango commented:
@b0661, just pushed back a version with the missing APIs.

Though, you should be aware that there will be more changes coming, as @galak is adding some changes on top of that and initial merge will be a stripped version of that with some addition of EDTSDeviceClass.

b0661 commented:
@erwango thanks for the info. I´m fine with having the API back.

b0661 commented:
@erwango 

What about get_device_id_by_label()?

erwango commented:
> What about get_device_id_by_label()?

Was renamed into get_device_id**s**_by_label, but does not make any sense. Back to get_device_id_by_label now.

galak committed:
scripts: edts: Add api to get device_ids based on compat

Added get_device_ids_by_compatible to EDTSConsumerMixin

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
scripts: edts: Add api to get device_id based on label

Added get_device_id_by_label to EDTSConsumerMixin

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
scripts: edts: Add api to get list of compatiables

Added get_compatibles to EDTSConsumerMixin

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
scripts: edts: Add api to get list of aliases

Added get_aliases to EDTSConsumerMixin

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
scripts: edts: Add api to get list of chosen

Added get_chosen to EDTSConsumerMixin

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
scripts: edts: Add api to get list of device types

Added get_device_types to EDTSConsumerMixin

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
scripts: edts: Add api to get list of controllers

Added get_controllers to EDTSConsumerMixin

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak commented:
I pushed an update that introduces EDTSDevice class and splits out the consumer updates into multiple commits.

@erwango can you look at fixing up your test script for the introduction of EDTSDevice

erwango committed:
scripts/dts: provide edts sanity check scripts

Validate generated etds.json against a reference edts file.
Script first checks edts structure then content of various elements
and finally devices parameters values.
This last step is done using a specific method for ref_edts
parameter extraction so we can verify behavior of
edtsdatabase class extraction function.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented:
@galak:

edtscheck.py updated, now using EDTSDevice methods.

b0661 commented:
@erwango, @galak 

I created an EDTS database variant with an integrated extraction function that can directly extract from the DTS compiled file. The consumer, provider and database layout is a copy of #9876. For codegen development I will use this variant of the EDTS database. My plan is to regularly synchronize the API and functionality to #9876. As I'm working on a copy variant (and not on #9876) there will be no clashed in case of incompatible API changes.

You may have a look at https://github.com/b0661/zephyr/commit/3b7c77f25654494d9c85fbcbd8ab580e6e1489a5

Major differences/ improvements to #9876 are:

  - the database now has an own extraction function that can be used instead of e.g. extract_dts_includes.  The extraction function follows the design of  the extract_dts_includes script and the additions that  were done in #9876. It is restructured and several globals are now classes and objects. All functionality of extract_dts_includes related to the generation of defines is not part of the database extract function. It's sole purpose is to fill the database directly from the compiled DTS file.
  - the database got itś own directory 'edtsdb' to structure all files related to the database.
  - The EDTSDevice class from #9876 was enhanced to allow devices to access the database they are taken from.
    
Mayor compatibility issues to #9876.
  - the consumer, provider API and the internal structure  of the databse is copied from #9876. API should be fully compatible.
  - the EDTSDevice class is copied from #9876. The  device API should be fully compatible except for       the unique id feature. Device name can be used instead.

galak commented:
Closing, as this should be superceded by PR #17660
